hypothesis-python/src/hypothesis/strategies/_internal/lazy.py

```diff
@@ -23,46 +24,52 @@
 )
 from hypothesis.strategies._internal.deferred import DeferredStrategy
 from hypothesis.strategies._internal.strategies import Ex, RecurT, SearchStrategy
+from hypothesis.utils.threading import ThreadLocal
 
 unwrap_cache: MutableMapping[SearchStrategy, SearchStrategy] = WeakKeyDictionary()
-unwrap_depth = 0
+# I don't think making this threadlocal is necessary since we also have
+# unwrap_strategies_lock, but the overhead is so small that I'm doing so for safety.
+threadlocal = ThreadLocal(unwrap_depth=0)
 
+# only let one thread write to unwrap_cache at once. We'd still like to keep the
+# cache global.
+unwrap_strategies_lock = RLock()
```

Comment:
Since we clear the cache when we're done unwrapping, and we have a lock protecting that, it seems like we're not meaningfully using a global cache here?  In which case I'd just make the `unwrap_cache` thread-local too.

Reply:
oh shoot right I didn't realize `unwrap_cache` only lives for the duration of the function. I had in mind that "cached strategy definitions should ideally still be global / shared across threads" when writing this